File: C/Tree/Tree_08.c
/*
===============================================================================
Program Name    : Binary Tree - Search Operation
Program Number  : 008
File Name       : Tree_08.c
Description     : This program demonstrates the Search() operation of a
                  Binary Search Tree (BST) using C. The Search() function
                  traverses the tree based on the Binary Search Tree
                  property and determines whether the specified element
                  is present or not.

Date            : August 2026
Language        : C
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Structure
2. Self-Referential Structure
3. Pointer
4. Binary Search Tree (BST)
5. Tree Traversal
6. Searching

Key Learning Points:
- Search an element in a BST.
- Traverse the tree efficiently.
- Apply the BST ordering property.
- Return search status using Boolean values.

===============================================================================
*/
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>

#pragma pack(1)
struct node
{
    int data;
    struct node *lchild;
    struct node *rchild;
};

typedef struct node NODE;
typedef struct node* PNODE;
typedef struct node** PPNODE;

// L    D   R
void Inorder(PNODE first)
{
    if(first != NULL)
    {
        Inorder(first -> lchild);
        printf("%d\n",first -> data);
        Inorder(first -> rchild);
    }
}

// D    L   R
void Preorder(PNODE first)
{
    if(first != NULL)
    {
        printf("%d\n",first -> data);
        Preorder(first -> lchild);
        Preorder(first -> rchild);
    }
}

// L    R   D
void Postorder(PNODE first)
{
    if(first != NULL)
    {
        Postorder(first -> lchild);
        Postorder(first -> rchild);
        printf("%d\n",first -> data);
    }
}

void Insert(PPNODE first, int iNo)
{
    PNODE newn = NULL;
    PNODE temp = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn -> data = iNo;
    newn -> lchild = NULL;
    newn -> rchild = NULL;

    if(*first == NULL)
    {
        *first = newn;
    }
    else
    {
        temp = *first;

        while(1)
        {
            if(iNo > temp -> data)
            {
                if(temp -> rchild == NULL)
                {
                    temp -> rchild = newn;
                    break;
                }
                temp = temp -> rchild;
            }
            else if(iNo < temp -> data)
            {
                if(temp -> lchild == NULL)
                {
                    temp -> lchild = newn;
                    break;
                }
                temp = temp -> lchild;
            }
            else if(iNo == temp -> data)
            {
                printf("Unable to Insert as Elements is Duplicate");
                free(newn);
                break;
            }
        }
    }
}

bool Search(PNODE first, int iNo)
{
    bool bFlag = false;

    while(first != NULL)
    {
        if(iNo == first -> data)
        {
            bFlag = true;
            break;
        }
        else if(iNo > first -> data)
        {
            first = first -> rchild;
        }
        else if(iNo < first -> data)
        {
            first = first -> lchild;
        }
    }
    return bFlag;
}

int Count(PNODE first)
{
    static int iCount = 0;
    

    if(first != NULL)
    {
        iCount++;
        Count(first -> lchild);
        Count(first -> rchild);
    }
    return iCount;
}

int main()
{
    PNODE head = NULL;
    int iRet = 0;

    Insert(&head,11);
    Insert(&head,5);
    Insert(&head,17);
    Insert(&head,21);
    Insert(&head,4);
    Insert(&head,7);
    Insert(&head,15);

    printf("Inorder Display: \n");
    Inorder(head);

    iRet = Count(head);
    printf("Number of Nodes are: %d\n",iRet);

    if(Search(head,17) == true)
    {
        printf("17 is present in BST");
    }
    else
    {
        printf("17 is not present in BST");
    }
    
    return 0;
}
/*
===============================================================================
Expected Output:
===============================================================================

Inorder Display:

4
5
7
11
15
17
21

Number of Nodes are: 7

17 is present in BST

===============================================================================
Memory Representation:
===============================================================================

                 11
               /    \
              5      17
             / \    /  \
            4   7  15  21


Searching for 17

11
 │
 ▼
17  ✓ Found

===============================================================================
Algorithm:
===============================================================================

1. Start from the Root node.
2. Compare the search element with the current node.
3. If equal, return TRUE.
4. If smaller, move to the left subtree.
5. If greater, move to the right subtree.
6. Repeat until the element is found or NULL is reached.
7. Return FALSE if the element is not present.

===============================================================================
Program Flow:
===============================================================================

            Start
               │
               ▼
      Start from Root
               │
               ▼
     Element == Current ?
        │             │
      Yes            No
       │              │
       ▼              ▼
 Return TRUE    Compare Value
                     │
           ┌─────────┴─────────┐
           ▼                   ▼
     Move Left           Move Right
           │                   │
           └─────────┬─────────┘
                     ▼
              NULL Reached?
                │        │
              Yes       No
               │         │
               ▼         │
          Return FALSE ◄─┘

===============================================================================
Time Complexity:
===============================================================================

Best Case    : O(log n)

Average Case : O(log n)

Worst Case   : O(n)

Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 008 demonstrates the Search() operation of a Binary Search Tree
using C. By utilizing the Binary Search Tree property, the program
efficiently locates the required element and returns whether it is
present or absent in the tree.

===============================================================================
*/

File: Java/Tree/Tree_08.java
/*
===============================================================================
Program Name    : Binary Search Tree - Count Leaf Nodes
Program Number  : 08
File Name       : Tree_08.java

Description     : This program demonstrates how to count the number of leaf
                  nodes in a Binary Search Tree using Java. A leaf node is a
                  node which does not have any left child or right child.
                  The program uses recursion to count all leaf nodes.

Date            : August 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Class
2. Object-Oriented Programming (OOP)
3. Self-Referential Class
4. Constructor
5. Binary Search Tree (BST)
6. Dynamic Memory Allocation
7. Recursion
8. Insertion Operation
9. Leaf Node
10. Function Overloading

Key Learning Points:
- Understand the concept of a leaf node in a Binary Search Tree.
- A leaf node does not have a left child or right child.
- Learn how recursion is used to count leaf nodes.
- Return 0 when the current node is NULL.
- Return 1 when a leaf node is found.
- Add the leaf count of the left and right subtrees.

===============================================================================
*/

class node
{
    public int data;
    public node lchild;
    public node rchild;
}

class Tree
{
    private node first;

    public Tree()
    {
        first = null;
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Insert
    // Description   : Insert element into Binary Search Tree
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Insert(int iNo)
    {
        node newn = null;
        node temp = null;

        newn = new node();

        newn.data = iNo;
        newn.lchild = null;
        newn.rchild = null;

        if(first == null)
        {
            first = newn;
        }
        else
        {
            temp = first;

            while(true)
            {
                if(iNo > temp.data)
                {
                    if(temp.rchild == null)
                    {
                        temp.rchild = newn;
                        break;
                    }

                    temp = temp.rchild;
                }
                else if(iNo < temp.data)
                {
                    if(temp.lchild == null)
                    {
                        temp.lchild = newn;
                        break;
                    }

                    temp = temp.lchild;
                }
                else
                {
                    System.out.println("Duplicate element not allowed");
                    break;
                }
            }
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : CountLeaf
    // Description   : Count leaf nodes in the tree
    //
    ///////////////////////////////////////////////////////////////////////////

    public int CountLeaf(node temp)
    {
        if(temp == null)
        {
            return 0;
        }

        if(temp.lchild == null && temp.rchild == null)
        {
            return 1;
        }

        return CountLeaf(temp.lchild) +
               CountLeaf(temp.rchild);
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : CountLeaf
    // Description   : Start counting leaf nodes from first
    //
    ///////////////////////////////////////////////////////////////////////////

    public int CountLeaf()
    {
        return CountLeaf(first);
    }
}

class Tree_08
{
    public static void main(String A[])
    {
        Tree tobj = new Tree();

        tobj.Insert(11);
        tobj.Insert(5);
        tobj.Insert(17);
        tobj.Insert(4);
        tobj.Insert(7);
        tobj.Insert(15);
        tobj.Insert(21);

        int iRet = tobj.CountLeaf();

        System.out.println("Number of Leaf Nodes : " + iRet);
    }
}

/*
===============================================================================
Expected Output:
===============================================================================

Number of Leaf Nodes : 4

===============================================================================
Memory Representation:
===============================================================================

                         11
                       /    \
                      5      17
                     / \    /  \
                    4   7  15  21

Leaf Nodes:

              4       7       15       21
              |       |        |        |
            Leaf     Leaf     Leaf     Leaf

Total Leaf Nodes = 4

===============================================================================
Leaf Node Identification:
===============================================================================

A leaf node is a node whose:

lchild = NULL
rchild = NULL

For the given Binary Search Tree:

Node 4  -> Leaf
Node 7  -> Leaf
Node 15 -> Leaf
Node 21 -> Leaf

Total = 4 Leaf Nodes

===============================================================================
Counting Process:
===============================================================================

Step 1 : Start from root node 11.
Step 2 : Visit the left subtree of 11.
Step 3 : Visit node 5.
Step 4 : Visit node 4. Node 4 is a leaf node, return 1.
Step 5 : Visit node 7. Node 7 is a leaf node, return 1.
Step 6 : Visit the right subtree of 11.
Step 7 : Visit node 17.
Step 8 : Visit node 15. Node 15 is a leaf node, return 1.
Step 9 : Visit node 21. Node 21 is a leaf node, return 1.
Step 10: Add all leaf node counts.

Total = 1 + 1 + 1 + 1
      = 4

===============================================================================
Node Structure:
===============================================================================

                    +-----------+-----------+-----------+
                    |   data    |  lchild   |  rchild   |
                    +-----------+-----------+-----------+
                          |           |           |
                        Value       Address     Address

Leaf Node Structure:

                    +-----------+-----------+-----------+
                    |   data    |   NULL    |   NULL    |
                    +-----------+-----------+-----------+

===============================================================================
Algorithm:
===============================================================================

1. Start the program.
2. Define the node class with data, lchild and rchild members.
3. Define the Tree class.
4. Initialize first to NULL using the constructor.
5. Define the Insert function.
6. Insert the given elements into the Binary Search Tree.
7. Define the recursive CountLeaf function.
8. Check whether the current node is NULL.
9. If the current node is NULL, return 0.
10. Check whether lchild and rchild are both NULL.
11. If both are NULL, return 1 because the node is a leaf node.
12. Recursively count leaf nodes in the left subtree.
13. Recursively count leaf nodes in the right subtree.
14. Add both counts and return the result.
15. Call CountLeaf starting from the first node.
16. Display the number of leaf nodes.
17. Stop the program.

===============================================================================
Program Flow:
===============================================================================

        Start
          |
          v
     Create Tree
          |
          v
   Insert Tree Elements
          |
          v
    Call CountLeaf()
          |
          v
      Start at Root
          |
          v
      Is Node NULL?
        /       \
      Yes        No
       |          |
       v          v
   Return 0   Is Leaf Node?
                /     \
              Yes      No
               |        |
               v        v
           Return 1  Count Left
                        |
                        v
                    Count Right
                        |
                        v
                    Add Counts
                        |
                        v
                  Return Result
                        |
                        v
                  Display Result
                        |
                        v
                       Stop

===============================================================================
Time Complexity:
===============================================================================

Insertion : O(log n) Average
Insertion : O(n) Worst Case
CountLeaf : O(n)

===============================================================================
Space Complexity:
===============================================================================

Tree Storage    : O(n)
Recursion Stack : O(h)

Where h is the height of the Binary Search Tree.

===============================================================================
Conclusion:
===============================================================================

Program 08 demonstrates how to count the number of leaf nodes in a Binary
Search Tree using Java. The recursive CountLeaf function checks every node
and identifies a leaf node when both lchild and rchild are NULL. For the
given Binary Search Tree, the total number of leaf nodes is 4.

===============================================================================
*/
